cpp/src/arrow/dataset/filter.cc

```diff
@@ -1013,14 +1013,17 @@ std::vector<std::string> FieldsInExpression(const ExpressionPtr& expr) {
 RecordBatchIterator ExpressionEvaluator::FilterBatches(RecordBatchIterator unfiltered,
                                                        ExpressionPtr filter,
                                                        MemoryPool* pool) {
-  auto filter_batches = [filter, pool, this](
-                            const std::shared_ptr<RecordBatch>& unfiltered,
-                            std::shared_ptr<RecordBatch>* filtered, bool* accept) {
-    ARROW_ASSIGN_OR_RAISE(auto selection, Evaluate(*filter, *unfiltered, pool));
-    ARROW_ASSIGN_OR_RAISE(*filtered, Filter(selection, unfiltered, pool));
-    // drop empty batches
-    *accept = (*filtered)->num_rows() > 0;
-    return Status::OK();
+  auto filter_batches = [filter, pool, this](std::shared_ptr<RecordBatch> unfiltered) {
+    auto filtered = Evaluate(*filter, *unfiltered, pool).Map([&](Datum selection) {
```

Comment:
The lambda passed to `Result::Map` is evaluated immediately, so there's no lifetime issue there. The `filter` is copied into `filter_batches`' closure, so its lifetime is guaranteed for the lifetime of the output iterator. The `this` capture might be subtler; conceivably you could create an evaluator on the stack then call this method. If you did and the evaluator were destroyed before iteration completed on the result, that would lead to undefined behavior. (The lifetime of evaluators is guaranteed by the Scanner interface; the above is very much an edge case)